CI: Unify code_checks whitespace checking

ShaharNaveh opened the pull request:
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

Unify test cases of #30467 #30708 #30737

Commit:
CI: Unify tests cases

alimcmaster1 commented:
For #30737. Is it possible to do this via regex? similar to the current code check for context manager:
https://github.com/pandas-dev/pandas/blob/master/ci/code_checks.sh#L157

MomIsBestFriend commented:
> For #30737. Is it possible to do this via regex? similar to the current code check for context manager:
> https://github.com/pandas-dev/pandas/blob/master/ci/code_checks.sh#L157

Yes, for simple cases like

```python
with pytest.raises(ValueError)
```
or (For false positives)
```python
with pytest.raises(ValueError, match="foo")
```


But I want a "global" test case.

So even if there's an edge case, for example:

```python
with pytest.raises(
                       ValueError,
                       # Some comment
)
```

or (For the case of false positives)

```python
with pytest.raises(
                       ValueError,
                       # Some comment
                       match="foo"
)
```


Regex is very fast, and for a reason, regex is not looking in the ```ast```. But it's hard for me to predict the future edge cases, and write regex to handle them.

So I wrote a global test case that looks in the ```ast```.

When I detect the occurrence of ```pytests.raises```, I brute-force all the tokens in the logical line until the end of the logical line, and if I find the occurrence of ```match``` as a argument (```token.NAME```) I consider this logical line as good.

My point is, regex is nice for "simple things" but when you have so many edge cases, I'd like to stick to checking it via the ```ast```.

P.S
If you're not convinced yet @alimcmaster1 I have some more edge cases that I couldn't found matching regex for them, If you succeed doing so I'd be more than happy to see the solution:)

alimcmaster1 commented on scripts/validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,256 @@
+    return is_failed
+
+
+def STC(source_path: str) -> Generator[Tuple[str, int, str], None, None]:
```
https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

Can we rename the methods to something more readable? Thanks!

MomIsBestFriend replied:
Yes, you're absolutely right

Commit:
Changed functions name

WillAyd reviewed (changes requested):
Any chance you can make the edits in place first without changing files? Typically easier to review that way

If you are moving and modifying things easier to do separately

datapythonista commented:
Agree with Will, do you mind leaving the name unchanged in this PR, even if it's not accurate, so we can review the new changes only. Then it's easy to just rename in a follow up, and get that merged quickly.

Just one early comment, I'm -1 on creating abbreviations like `STD`. I think it's worth having proper names that can be directly understood, even if they are longer.

MomIsBestFriend commented:
> Any chance you can make the edits in place first without changing files?

Not sure what you mean by that @WillAyd , can you please be more specific?

datapythonista commented:
Check my last comment

MomIsBestFriend commented:
> Check my last comment

So just change the function names, back to the __very__ original?

datapythonista commented:
No, sorry. The file name, so the diff will show what you changed in that file.

Commit:
Renamed file name, back to the original

Commit:
STY: inconsistent linebreaks

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Made the code check "ID" to be more verbose

as @datapythonista suggested in https://github.com/pandas-dev/pandas/pull/30755#issuecomment-571533520

Commit:
Added ".py" in ci/code_check.sh to the script call

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+        "strings_to_concatenate": strings_to_concatenate,
+        "strings_with_wrong_placed_space": strings_with_wrong_placed_space,
+        "bare_pytest_raises": bare_pytest_raises,
+    }
```
This is not needed, you can get the function from `globals()` (e.g. `globals()[args.id]`)

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+        help="Output format of the error message.",
+    )
+    parser.add_argument(
+        "--id", "-i", choices=FUNCTIONS_MAP.keys(), help="Test case to check."
```
I don't think `id` is very clear, `validation-type` or something similar is more descriptive. I'd also avoid calling them "test cases", since that is confusing with unit testing terminology.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,51 +1,39 @@
-Black is not considering this as an
-issue (see issue https://github.com/psf/black/issues/1051),
-so we are checking it here.
+Unwanted patterns test cases.
```
Can you give some context here? Like why this script exists (why this stuff is not in `ci/code_checks.sh` directly like the rest.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
     """
+    MSG: str = (
+        "String unnecessarily split in two by black. Please merge them manually."
+    )
```
I'd personally have the string directly where it's used, but up to you, it's just a preference.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+
+    >>> foo = (
+    ...        "bar"
+    ...       " baz"
```
This indentation is different than the previous, and makes it more difficult to compare. Also, if instead of `foo`, `bar`... you explain what's going on, the same example can be clearer. e.g.:
```python
rule = (
    "We want the space at the end of lines, "
    "not at the beginning"
)
```

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+        Source file path.
+    line_number : int
+        Line number of unconcatenated string.
+    MSG : str
```
using uppercase here feel a bit confusing, even if it somehow makes sense.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+            first_token[0] == third_token[0] == token.STRING
+            and second_token[0] == token.NL
+        ):
+            # Means we are in a block of concated string
```
May be just a preference, but I think this is more clear if it's placed before the if, not inside.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+
+    for first_token, second_token, third_token in zip(tokens, tokens[1:], tokens[2:]):
+        if (
+            first_token[0] == third_token[0] == token.STRING
```
I think `0` means `TYPE` here, can you define a constant for those at the beginning of the file, so this is easier to understand? Same for the rest.

MomIsBestFriend replied:
> I think `0` means `TYPE` here, can you define a constant for those at the beginning of the file, so this is easier to understand? Same for the rest.

Not sure what you mean, on defining a constant here? can you be more explicit?

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+            second_string = third_token[1][1:-1]
+
+            if (not first_string.endswith(" ")) and (second_string.startswith(" ")):
+                yield source_path, third_token[2][0], MSG
```
Don't we want to capture things like:
```python
msg = (
    "Space at the end "
    " and at the beginning."
)
If I'm not wrong this won't raise an error.

Also, the parenthesis aren't needed.

Finally, as commented in the other PR, I'd just validate when there is a space, and will ignore when there are several together.

MomIsBestFriend replied:
> If I'm not wrong this won't raise an error.

You are not wrong, will fix.

---

> Also, the parenthesis aren't needed.

I worked on the past two days about this exact issue, will open a PR soon.

MomIsBestFriend replied:
> Don't we want to capture things like:
> 
> ```python
> msg = (
>     "Space at the end "
>     " and at the beginning."
> ```

I think this deserves its own test case, because the "error" message should be different from ```"String has a space at the beginning instead of the end of the previous string."```

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+    """
+    Test Case for bare pytest raises.
+
+    For example:
```
```suggestion
    For example, this is wrong:
```

MomIsBestFriend replied:
Applied.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+    >>> with pytest.raise(ValueError):
+    ...     # Some code that raises ValueError
+
+    Instead of:
```
```suggestion
    And this is what we want instead:
```

Otherwise, it's difficult to understand which is the good one or not (I'd guess it's the opposite if I didn't know.

MomIsBestFriend replied:
Applied.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+        tokens: List = list(tokenize.generate_tokens(file_name.readline))
+
+    for counter, current_token in enumerate(tokens, start=1):
+        if current_token[0] == token.NAME and current_token[1] == "raises":
```
Small preference, but I'd negate this condition and have a `continue` for it. So the rest of the code doesn't need to be nested. I think there are too many levels of indentation here.

MomIsBestFriend replied:
I don't see where the ```continue``` will be place in order to have less code indented.

Can you help me out?

datapythonista replied:
See these two cases:
```python
for item in items:
    if condition:
        do_stuff()
```
and
```python
for item in items:
    if condition:
        continue
    do_stuff()
```
If in the place of `do_stuff()` you've got lots of lines, the second is probably more readable, because you don't have a big block indented many lines, that is difficult to know exactly at which level it is.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -102,28 +106,157 @@ def strings_to_concatenate(source_path: str) -> Generator[Tuple[str, int], None,
+            for next_token in tokens[counter:]:
+                if next_token[0] == token.NAME and next_token[1] == "match":
+                    break
+                if next_token[0] == token.NEWLINE:
```
Does `NEWLINE` means the colon here? If it is, I'd add a comment, since it's a bit confusing. If it's not, I think we're not validating properly things like:
```python
pytest.raises(ValueError
              match="whatever"):
   pass

MomIsBestFriend replied:
```token.NEWLINE``` doesn't necesarly means ```:``` but in many cases it does, ```token.NEWLINE``` represents the end of a logical line, so for example for the line of

```python
msg = "my message"
```

the ```token.NEWLINE```

will be "placed" here:

```python
msg = "my message"token.NEWLINE
```

and for strings with multi lines for example:

```python
msg = (
    "foo "
     "bar "
     "baz"
)token.NEWLINE
```

there's also a ```token.NL``` which represents a new line (```\n```) but it's different from ```token.NEWLINE``` because it's "stupider", because it can't tell the diffrence between a new line a new *logical* line.

If I have to put ```token.NL``` next to some python code it will probably will look like this:

```python
msg = (token.NL
    "foo "token.NL
     "bar "token.NL
     "baz"token.NL
)token.NEWLINE
```

datapythonista replied:
Great, can you add a short comment that makes clear that `NEWLINE` is not a new line in the source file please. This is misleading when it's read.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,51 +1,39 @@
 
 
-def main(source_path: str, output_format: str) -> bool:
+def main(
```
Not sure if it's just my own convention, but I'd say you usually sort functions in the opposite order. So the last one is the `if __name__...` block, the previous is `main`, and the first functions are the ones that don't call anything from this function.

The idea is that when a function is being defined, the functions that it uses are already defined (in compiled languages things can fail otherwise).

datapythonista reviewed (commented):
Nice stuff. Some comments, but the general idea looks good.

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Replaced the dictonary with 'globals()'

Commit:
Changed from "id" to "validation-type"

Commit:
Better explained (I hope) why this file exists

Commit:
Yielding the string directly instead of assigning it to a variable

Commit:
Fixed indentation in docs

Commit:
Imporved docstring for spaces test case, as datapythonista suggested

Commit:
Removed uppercase "MSG" from docs and replaced it with lowercase "msg"

Commit:
Fixed comment placement

Commit:
Applied datapythonista suggestions

Commit:
Moved "main" function to the bottom (Above 'if __name__ == "__main__"')

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Imporved test case in 'strings_with_wrong_placed_space' validation type

pep8speaks commented:
Hello @MomIsBestFriend! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`scripts/validate_string_concatenation.py`](https://github.com/pandas-dev/pandas/blob/b60f5ee63857211f59f6966090a4f786bee709ed/scripts/validate_string_concatenation.py):

> [Line 125:89](https://github.com/pandas-dev/pandas/blob/b60f5ee63857211f59f6966090a4f786bee709ed/scripts/validate_string_concatenation.py#L125): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (89 > 88 characters)

Commit:
Added more logic to 'strings_with_wrong_placed_space'

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,51 +1,217 @@
+                (not is_first_ends_single)
+                and is_second_starts_single
+                and (not is_second_starts_double)
+            ) and (not is_first_ends_newline):
```
I don't think all this part is very easy to read. Probably worth creating functions for the conditions. Something like:
```
def single_space_in_wrong_place(line1, line2):
    return line1[-1] != " " and line2[0] == " " and line2[1] != " "
```

MomIsBestFriend replied:
What should the function return?  a tuple of (bool, str)

bool - indicating if the two strings passed/failed
str - the correct error message (ref https://github.com/pandas-dev/pandas/pull/30755#discussion_r365288596)

datapythonista replied:
I think just the bool will make the code better. See the example.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -87,43 +251,37 @@ def main(source_path: str, output_format: str) -> bool:
+        f.__name__
+        for f in globals().values()
+        if type(f) == type(main) and f.__name__ != "main"
+    ]
```
I think it's probably better to have the list of functions explicit. So, it's clearer, and we have control on what can be called.

datapythonista reviewed (commented):
Thanks for the updates, nice work.

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Added some typings

Commit:
Defining 'TYPE' as constant

Commit:
Added 'continue' for code block

Commit:
Added comment about token.NEWLINE

Commit:
Applied datapythonsita code suggestion

Commit:
Replaced available_validation_types with predefined list of available validation types

Commit:
Make the validation type parameter to be required

Commit:
Added 'VALUE' constant

Commit:
Fixed grammer mistake (I think)

Commit:
Removed complex if statement, Replaced with  a function

MomIsBestFriend commented on scripts/validate_string_concatenation.py:
```diff
@@ -111,6 +111,30 @@ def strings_with_wrong_placed_space(
         Explenation of the error.
     """
+
+    def idk_how_to_call_this_function(first_line: str, second_line: str) -> bool:
```
I really don't know how to call this :smile:

datapythonista replied:
Naming things and cache invalidation are the two most challenging problems in CS. :)

Sorry, the expression is too difficult to understand. Do you mind using something like:
```python
if first_line.endswith(r'\n'):
   return False
if second_line.ends_with('  '):
    return False
...
```
Happy to help find a name when I understand what's this doing. Also, may be worth adding an example sections, since this function seems somehow complex.

You have couple of typos in the docsting, and I think `r'\n'` is more readable than `'\\n'`.

Thanks!

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Removed useless parenthasis

Commit:
Fix tests case for 'strings_with_wrong_placed_space'

Commit:
Removed constants and replaced with values from the generated named tuple

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Fixed CI-test for 'strings_with_wrong_placed_space'

Commit:
Activating "strings_with_wrong_placed_space" in the CI

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Refactored the function that check the unwanted patterns of the wrong placed spaces

Commit:
Added docs

Commit:
Changed function name

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Rearanged functions in alphabetical order

Commit:
Renamed strings_with_wrong_placed_space --> strings_with_wrong_placed_whitespace

Commit:
Fixed bug where not report wrong placed whitespace if string had a prefix of a string literal

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Changed validation-type call name in the CI as well

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,51 +1,313 @@
+                yield (
+                    source_path,
+                    current_token.start[0],
+                    "Bare pytests raise have been found.",
```
Probably worth adding what to do, I think this message can be a bit cryptic for beginners.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,51 +1,313 @@
+
+        Examples
+        --------
+
```
Better remove this blank line.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,51 +1,313 @@
+        >>> get_prefix_len('rf"Hello world"')
+        2
+        """
+        for counter, character in enumerate(line):
```
I don't think you need to loop, you can use:
```python
>>> my_string = "rf'foo'"
>>> my_string.index("'")
2

```

MomIsBestFriend replied:
The two reasons I'm using a loop are:

* I don't know the length of the prefix (If even has a prefix)

* I don't know the correct character that the string is wrapped in (I'll explain):

I can't predict what the character will be, means that I can't know if the string is represented like this:

```python
my_str = "'Hello world'"
```

or like that:

```python
my_str = '"Hello world"'
```

datapythonista replied:
I see, but I still think that something like:
```python
min(line.find(quote) for quote in ('"', "'") if line.find(quote) >= 0)
```
should return what you want without having to iterate over the string.

MomIsBestFriend replied:
__WoW__.
you just blew me away, that's super smart!
I am for real speechless, that's so smart.

datapythonista commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,51 +1,313 @@
+                return counter
+        return 0
+
+    def whitespace_validation(first_line: str, second_line: str) -> bool:
```
Since this returns True/False, I think it can make things clearer if the name of the function specifies what's True and what's False. For example `has_wrong_whitespaces`.

datapythonista reviewed (approved):
Really nice work. Couple of suggestions that I think should make things clear, but looks great. Thanks @MomIsBestFriend

Commit:
Eleborate more about bare_pytest_raises error message

Commit:
Removed empty line below 'Examples'

Commit:
Renamed "whitespace_validation" function to --> "has_wrong_whitespaces"

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Removed useless function, with a brilliant idea

datapythonista reviewed (approved):
Thanks for the update, and for the kind words. Looks perfect.

@WillAyd do you mind reviewing again please?

WillAyd commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,51 +1,293 @@
+            )
+
+
+def strings_with_wrong_placed_whitespace(
```
Are there tests for these functions?

MomIsBestFriend replied:
No there are no tests for these functions.

I have no clue how to tests these, If you can give me a minimal starting point (how to test one of these functions, or an abstract idea on how to test them) I believe I will be good to go and test every other function and expand the testing even further, I just really don't know where to start.

datapythonista replied:
You can add a new file in `scripts/tests/`, check the docstrings one for reference.

Then you can start with a simple test for each function. A function with an assert, something like this for your functions:
```python
def test_sum_works():
    assert sum(1, 2) == 3
```

Then we can review or add comments, or if you feel confident, you can directly test several uses cases, similar to the examples you wrote for the wrong placed whitespaces.

MomIsBestFriend replied:
@datapythonista ty for the quick reply :)

Do you mind if I will create in addition to the test file, another file (possibly 3) that will contain some data for the tests? because the functions only rake one parameter and that's a file path.

datapythonista replied:
That should be ok, but probably better to use `io.StringIO`. Making the functions accept a file descriptor instead of a file path is not a bad idea either. I think it's common, and it can be useful not only for the tests, but for validating content that is not in the file system, but in a url, or somewhere else.

In any case, feel free to implement whatever you think it makes more sense. But I think it's better to write unit tests, testing minimalistic and specific cases, than have files on disk.

WillAyd replied:
I don't think should need to create additional files. Should be able to use https://github.com/pandas-dev/pandas/blob/6b87716e1416061b4bf0109a89caf962acdf8c12/scripts/tests/test_validate_docstrings.py#L16 as an example but create another class that has methods exhibiting the behavior you are trying to correct here instead of docstrings

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Refactored the validation-type functions, to get a file object

Previously the validation-type functions were getting a string
that was the file/directory path

MomIsBestFriend commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,129 +1,357 @@
 
 
-def main(source_path: str, output_format: str) -> bool:
+def bare_pytest_raises(file_obj: IO) -> Generator[Tuple[int, str], None, None]:
```
Can ```file_obj``` be typed as ```IO[str]``` or ```IO[bytes]``` ?

or maybe ```IO[AnyStr]``` ?

MomIsBestFriend commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,129 +1,357 @@
+    Parameters
+    ----------
+    file_obj : IO
+        File object to operate on.
```
Really not sure that this description is good.

Any thoughts?

datapythonista replied:
Could be something like "File-like object containing the Python code to validate."

MomIsBestFriend commented on scripts/validate_string_concatenation.py:
```diff
@@ -295,35 +281,42 @@ def main(
-                    source_path=source_path, line_number=line_number, msg=msg
+        file_path: str = source_path
+        with open(file_path, "r") as file_obj:
+            file_desc = file_obj.readline
```
Really not sure that ```file_desc``` is a good/correct name for this.

Any thoughts?

Commit:
Moved the "readline" to the function itself.

In order to call functions with io.StringIO

Commit:
Added some sample tests

Commit:
Changed functions names in the sample test

MomIsBestFriend commented:
@datapythonista @WillAyd 

I added a "sample" test file, just for you to tell me if I'm on the right track.

Do you have any thoughts/suggestions?

Commit:
Changed the names of the test functions as well

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,77 @@
+import validate_string_concatenation as validate_unwanted_patterns
+
+
+class GoodBarePytestRaises:
```
You don't need this class, or the Bad one, that's something that makes sense for the docstings (we have a class with methods, so we can extract the docstrings from them). But not here, just define the values in the tests when you need them.

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,77 @@
+        )
+
+
+class TestGoodBarePytestRaises:
```
This is ok, but I think the class doesn't add value in this case, just have the tests as functions.

MomIsBestFriend replied:
I am planning to test "strings_to_concatenate" and "strings_with_wrong_placed_whitespace", this is just a sample file, because I'm not sure what I do here is good or not.

Should I keep the class then or not?

datapythonista replied:
Using classes to group the tests of each function sounds good. Up to you.

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,77 @@
+class TestGoodBarePytestRaises:
+    def test_good_normal_pytest_raises(self):
+        result = validate_unwanted_patterns.bare_pytest_raises(
+            GoodBarePytestRaises.normal_pytest_raises(self)
```
Sorry the tests I told you to use as reference were too complicated. This is actually easier. Something like this is probably what we want:
```python
def test_bare_raises():
    fd = io.StringIO("""with pytest.raises(ValueError):
        raise ValueError("foo")""")
    result = list(validate_unwanted_patterns.bare_pytest_raises(fd))
    expected = [(1, "Bare pytests raise have been found. Please pass in the argument 'match' as well the exception.")]
    assert result == expected
```
The idea is that when our test suite executes, it'll call this function. This function will call the one you implemented with a known input, and this function will raise an assertion error, if your function is not returning what it's expected to return.

It's also great documentation to understand what your function is expected to receive as an input, and how the output looks like.

MomIsBestFriend replied:
@datapythonista Is it a good idea to not have the error message hard coded into the tests, but instead defining constants for each error message, in ```scripts/validate_unwanted_patterns```. ?

So if for some reason we want to change the error message we won't have to change it in many placed, and only in one place.

datapythonista replied:
I see your point, but I don't think that's common, and I wouldn't do that.

Having repetition in the tests is good. If you are reporting the wrong message in one exception, and you test that you're using the same constant you report, your test will pass, and the message will be incorrect. Better to keep things simple and explicit.

MomIsBestFriend replied:
Thank you for the explanation.

Commit:
Refactored tests

Commit:
Added tests to TestStringsToConcatenate

Commit:
Fixed typing

Commit:
Added tests for TestStringsWithWrongPlacedWhitespace

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

MomIsBestFriend commented:
@datapythonista Would you recommend "Squashing" some of these tests using ```pytest.mark.parameterize```?

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,215 @@
+        expected = []
+        assert result == expected
+
+    def test_strings_with_wrong_placed_whitespace_first_line_starts_double_sapce(self):
```
Typo: space

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,215 @@
+    def test_strings_to_concatenate(self):
+        fd = io.StringIO(
+            """
+msg = ("bar " "baz")
```
Can't this be a single line? (all in the same line as the `fd = ... `)

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,215 @@
+class TestBarePytestRaises:
+    def test_bare_pytest_raises(self):
+        fd = io.StringIO(
+            """
```
May be using dedent makes the content of the files more readable?

MomIsBestFriend replied:
I got to be honest, I don't see how it adds more readability here, I know this was suggested for the code in b2d2d62 but in 796550e I don't think this is the case (IMO). can you please "re-review" (re-think) for the case in 796550e?

datapythonista reviewed (commented):
Good work with the tests. Yes, it's a good idea to parameterize the repetitive tests.

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Changed paramerter description of "file_obj"

as @datapythonista suggested here
https://github.com/pandas-dev/pandas/pull/30755#discussion_r368203643

Commit:
Fixed typo

Ref https://github.com/pandas-dev/pandas/pull/30755#discussion_r368203493

Commit:
Refactored test to use "pytest.mark.parametrize"

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Added edge cases tests "for bare_pytest_raises"

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

MomIsBestFriend commented:
gentle ping @datapythonista

Commit:
Type annotations

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,443 @@
+    )
+    def test_bad_pytest_raises(self, fd, expected):
+        result = list(validate_unwanted_patterns.bare_pytest_raises(fd))
+        assert result == expected
```
Since this is the same as the previous, better to just have just  one test, merging the parameters of both. And you don't need the class anymore, since there is a single test in it.

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,443 @@
+            ),
+        ],
+    )
+    def test_strings_to_concatenate(self, fd, expected):
```
Same. No need for the class.

datapythonista commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,443 @@
+            ),
+        ],
+    )
+    def test_excluded_strings_with_wrong_placed_whitespace(self, fd, expected):
```
Same as before, merge both tests, and get rid of the class.

datapythonista reviewed (commented):
Looks great, thanks for working on this.

Just few final comments, to make the code a bit clearer, but almost ready.

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Making a function for getting the literal string prefix length

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r370911758

Commit:
Calling io.StringIO inside the bare_pytests_raises tests

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r370911991

Commit:
Calling io.StringIO inside the strings_to_concatenate and strings_with_wrong_placed_whitespace tests

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r370911991

Commit:
Changed comment to something more clear

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r370912125

Commit:
Unify bare_pytest_raises tests

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r370912235

Commit:
Unify strings_to_concatenate and strings_with_wrong_placed_whitespace tests

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r370912262
https://github.com/pandas-dev/pandas/pull/30755#discussion_r370912302

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

MomIsBestFriend commented:
ping @datapythonista

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

jreback commented:
@WillAyd over to you

WillAyd commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,423 @@
+        ),
+    ],
+)
+def test_pytest_raises(data, expected):
```
Can you split this into two tests where one is acceptable values and the other is a `_raises` test?

I'm hoping that will help limit the parametrization since you won't need to supply the error message if you do things that way. Right now the parametrization is pretty tough to read (partially due to black formatting rules)

WillAyd commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,423 @@
+            'msg = ("bar " "baz")',
+            [
+                (
+                    1,
```
IIUC `expected` is always the same, so move to body of test instead of parametrizing

MomIsBestFriend replied:
Actually ```expected``` is not always the same, One is checking that the check is detecting lines, and the other check that it also check the number of occurrences in each line.

---

For example if you have this:

```python
msg = "foo " "bar"
```

the check will output:
```
path/to/file:1 String unnecessarily split in two....
```

and for this:

```python
msg = "foo " "bar " "baz"
```

it will output:
```
path/to/file:1 String unnecessarily split in two....
path/to/file:1 String unnecessarily split in two....
```

WillAyd commented on scripts/tests/test_validate_unwanted_patterns.py:
```diff
@@ -0,0 +1,423 @@
+        ),
+    ],
+)
+def test_strings_with_wrong_placed_whitespace(data, expected):
```
Same comment; if you can split into acceptable patterns versus failing patterns will make much more readable

WillAyd commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,129 +1,377 @@
-from typing import Generator, List, Tuple
+from typing import IO, Callable, Generator, List, Tuple
+
+FILE_EXTENSIONS_TO_CHECK: Tuple[str, ...] = (".py", ".pyx", ".pyx.ini", ".pxd")
```
```suggestion
FILE_EXTENSIONS_TO_CHECK: Tuple[str, ...] = (".py", ".pyx", ".pxi.in", ".pxd")
```

MomIsBestFriend replied:
@WillAyd This apparently breaks the script.

WillAyd commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,129 +1,377 @@
+FILE_EXTENSIONS_TO_CHECK: Tuple[str, ...] = (".py", ".pyx", ".pyx.ini", ".pxd")
+
+
+def _get_literal_string_prefix_len(token_string: str) -> int:
```
Hmm what's the point of this function? Does it only ever return 0 or 1?

MomIsBestFriend replied:
#### Short version:
The point of this function is to get the prefix length from string literals, so I can know where the actual string starts.

---

#### Long version:

The received named tuple from generating the tokens, contains the "line" attribute, which returns the whole line of code as if it was a string for example for 

```python
value = 42
```

The received "line" value is: ```'value = 42'``` (note the wrapping single quotes).

The problem is that I don't know if the string will be wrapped with single or double quotes, so I might get this (as the value of "line"):

```python
"f'My age is {age}'"
```

or this:

```python
'f"My age is {age}"'
```

The function may return:

* 0 - if there is no string prefix

* 1 - if the string prefix is just one character (e.g ```f```, ```r```)

* 2 - if the string prefix is two letters (e.g ```fr```)

---

The function is used to determine where to start cutting the received string, in order to check only the "raw" string.

WillAyd commented on scripts/validate_string_concatenation.py:
```diff
@@ -1,129 +1,377 @@
+        return 0
+
+
+def bare_pytest_raises(file_obj: IO[str]) -> Generator[Tuple[int, str], None, None]:
```
```suggestion
def bare_pytest_raises(file_obj: IO[str]) -> Iterable[Tuple[int, str]]:
```

WillAyd reviewed (changes requested):
@MomIsBestFriend so have to admit I'm a little confused on purpose of this PR. Thought we were just moving things but this adds new tests instead right?

If that's true I think just need to simplify things a bit

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Seperated test case

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r377162037

Commit:
Splitted test cases

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r377163380

ShaharNaveh committed:
Update scripts/validate_string_concatenation.py

Co-Authored-By: William Ayd <william.ayd@icloud.com>

Commit:
Typing change as @WillAyd suggested

REF:
https://github.com/pandas-dev/pandas/pull/30755#discussion_r377165490

Commit:
Reverted a2a12fc39417721c5722e0fb9fbe17aff698e556

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

jbrockmendel commented on ci/code_checks.sh:
```diff
@@ -102,9 +102,17 @@ if [[ -z "$CHECK" || "$CHECK" == "lint" ]]; then
     MSG='Check for use of not concatenated strings' ; echo $MSG
     if [[ "$GITHUB_ACTIONS" == "true" ]]; then
-        $BASE_DIR/scripts/validate_string_concatenation.py --format="[error]{source_path}:{line_number}:{msg}" .
+        $BASE_DIR/scripts/validate_string_concatenation.py --validation-type="strings_to_concatenate" --format="##[error]{source_path}:{line_number}:{msg}" .
```
out of curiosity, what does the `##` here mean?

MomIsBestFriend replied:
Full disclosure I have no clue, I asked @datapythonista for the output format that it should be outputted when running on ```github-actions```.

datapythonista replied:
Just a convention in GitHub Actions. Any row in the output that starts with `##[error]` is highlighted as an error, and it's easier to find in the logs.

jbrockmendel commented:
This is... a lot.  It'd be great if one of the upstream tools (flake8/black/...) did this, but I'm not sure it belongs here.  i.e. if I had to maintain this myself, I would just learn to be OK with the downsides of regexps.

jbrockmendel commented:
updated the title for clarity, feel free to revert if I misunderstood

MomIsBestFriend commented:
@jbrockmendel The new title is correct, but not fully correct, this PR is basically merging the test cases introduced in #30467 #30708 #30737.

And it's also adding to the CI checks the test case introduced in #30708

I can't think of a better name for this PR unfortunately.

MomIsBestFriend commented:
> This is... a lot. It'd be great if one of the upstream tools (flake8/black/...) did this, but I'm not sure it belongs here. i.e. if I had to maintain this myself, I would just learn to be OK with the downsides of regexps.

@jbrockmendel I can see ```flake8``` __maybe__ adopting the ```unconcatnated strings```, but for the rest of the tests cases I can't see why a linter would adopt those, as it's a very specific ```pandas``` code style/nitpick, maybe a very opinionated linters such as ```pylint``` would use those.

(And do you know by any chance someone from ```flake8``` to ask if they would use it?
And I would love to see how they do it, I tried doing this code checks using the ```ast``` module, couldn't figure it out :/)

And about the regex, I completely understand why this PR seems like overkill when a simple regex can do the trick, for the most part. but isn't that the fact why we are using tools like ```black``` an ```isort```? so there will no questions if the code style right or not, there is always one correct answer and it is what the linter says, when we disagree with the decision of the linter for example in https://github.com/psf/black/issues/1051 we just create our own check (If it's not upstreamed, we do it here).

datapythonista commented:
I agree having these checks in the pandas code base is not ideal. But I'd move forward with this PR, get the validation working better (we already have the validation, this is just improving and refactoring what we've got). And later on, we can consider moving this out of pandas, as we did with the docstrings.

@MomIsBestFriend can you merge master and see if the CI is green please?

jbrockmendel commented:
I'm completely happy deferring to @datapythonista on this.  Thanks @MomIsBestFriend for taking this on.

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

MomIsBestFriend commented:
Restarting azure

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

MomIsBestFriend commented:
ping @datapythonista

datapythonista reviewed (approved):
Thanks @MomIsBestFriend 

@WillAyd I think your comments were addressed, can you have a look and merge if you're happy.

WillAyd commented:
I will relook in a few days

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

Commit:
Merge remote-tracking branch 'upstream/master' into CI-unify-boilerplate

ShaharNaveh committed:
Merge branch 'master' into CI-unify-boilerplate

datapythonista commented:
@MomIsBestFriend sorry we've been slow with this. Do you mind merging master once more please? I'll merge this once CI is green.

MomIsBestFriend commented:
@datapythonista I just merged master, I'll wait for another commit to be merged, before I can merge master again.

MomIsBestFriend commented:
ping @datapythonista :)

datapythonista commented:
Thanks @MomIsBestFriend, great job.

Can you follow up renaming the file please?
